src/Std/Sync/Channel.lean

```diff
@@ -1,137 +1,720 @@
+private structure Zero (α : Type) where
+  state : Mutex (Zero.State α)
+
+namespace Zero
+
+private def new : BaseIO (Zero α) := do
+  return {
+    state := ← Mutex.new {
+      producers := ∅
+      consumers := ∅
+      closed := false
+    }
+  }
+
+/--
+Precondition: The channel must not be closed.
 -/
-def Channel.recv? (ch : Channel α) : BaseIO (Task (Option α)) :=
-  ch.atomically do
+private def trySend' (v : α) : AtomicT (Zero.State α) BaseIO Bool := do
+  let st ← get
+  if let some (consumer, consumers) := st.consumers.dequeue? then
+    consumer.resolve (some v)
+    set { st with consumers }
+    return true
+  else
+    return false
+
+private def trySend (ch : Zero α) (v : α) : BaseIO Bool := do
+  ch.state.atomically do
+    if (← get).closed then
+      return false
+    else
+      trySend' v
+
+private def send (ch : Zero α) (v : α) : BaseIO (Task (Except Error Unit)) := do
+  ch.state.atomically do
+    if (← get).closed then
+      return .pure <| .error .closed
+    else if ← trySend' v then
+      return .pure <| .ok ()
+    else
+      let promise ← IO.Promise.new
+      modify fun st => { st with producers := st.producers.enqueue (v, promise) }
+      return promise.result?.map (sync := true)
+        fun
+          | none | some false => .error .closed
+          | some true => .ok ()
+
+private def close (ch : Zero α) : EIO Error Unit := do
+  ch.state.atomically do
+    let st ← get
+    if st.closed then throw .alreadyClosed
+    for consumer in st.consumers.toArray do consumer.resolve none
+    set { st with consumers := ∅, closed := true }
+    return ()
+
+private def isClosed (ch : Zero α) : BaseIO Bool :=
+  ch.state.atomically do
+    return (← get).closed
+
+private def tryRecv' : AtomicT (Zero.State α) BaseIO (Option α) := do
+  let st ← get
+  if let some ((val, promise), producers) := st.producers.dequeue? then
+    set { st with producers }
+    promise.resolve true
+    return some val
+  else
+    return none
+
+private def tryRecv (ch : Zero α) : BaseIO (Option α) := do
+  ch.state.atomically do
+    tryRecv'
+
+private def recv (ch : Zero α) : BaseIO (Task (Option α)) := do
+  ch.state.atomically do
     let st ← get
-    if let some (a, values) := st.values.dequeue? then
-      set { st with values }
-      return .pure a
+    if let some val ← tryRecv' then
+      return .pure <| some val
     else if !st.closed then
       let promise ← IO.Promise.new
       set { st with consumers := st.consumers.enqueue promise }
       return promise.result?.map (sync := true) (·.bind id)
     else
+      return .pure <| none
+
+end Zero
+
+/--
+The central state structure for a bounded channel, maintains the following invariants:
+1. `0 < capacity`
+2. `0 < bufCount → consumers = ∅`
+3. `bufCount < capacity → producers = ∅`
+4. `producers = ∅ ∨ consumers = ∅`, implied by 1, 2 and 3.
+5. `bufCount` corresponds to the amount of slots in `buf` that are `some`.
+6. `closed = true → producers = ∅ ∧ consumers = ∅`
```

Comment:
This invariant seems to be incorrect, as demonstrated by the following code:
```lean
def myTest : IO Unit := do
  let chan : ClosableChannel Nat ← ClosableChannel.new (some 1)
  let t ← chan.send 1
  chan.close
  let ret ← IO.wait t
  IO.println s!"{ret}"
  let v ← chan.tryRecv
  IO.println s!"{v}"

#eval myTest
```
The output is this:
```lean
ok: ()
(some 1)
```
The code behaves according to the spec of `close` here, so I think the invariant needs to be changed, not the code.

Furthermore, below you write "The `IO.Promise` will be resolved to `false` if the channel closes." which also seems to be incorrect.